### DISCUSSION TITLE
Get corrent environment variable with fish as default shell

### DISCUSSION THREAD

**Discussion by lakkiy** [d12frosted/homebrew-emacs-plus#418] [Q&A] [answered]
After zinit repo has been deleted I had switch to fish from zsh and set it as default shell on macOS 12.0.1, but when I run `genenv SHELL` on emacs-plus@28 I get `/bin/zsh`. 

I use emacs to write Golang program, and run Golang's `os.Getenv()` also can't get environment variable which I write this into my `config.fish` with `set -gx foo bar`.

I also install [exec-path-from-shell](https://github.com/purcell/exec-path-from-shell) but I have to set `exec-path-from-shell-shell-name` to `/usr/local/bin/fish` and it can read `PATH` into emacs, but other environment variable much run `exec-path-from-shell-copy-env` manually or Golang's `os.Getenv()` won't get this environment variable which means I can't run my program correctly on emacs-plus@28.

I also test this on [emacsmacport](https://github.com/railwaycat/homebrew-emacsmacport) and it can get correct `SHELL` which is `/usr/local/bin/fish`, and I don't need to set `exec-path-from-shell-name` manually cause it can get correct `SHELL`, and Golang's `os.Getenv()` works too.

**Comment by d12frosted** [comment]
Hey @404cn,

Environment variables is always a burden on macos and is a constant topic of confusion and troubles. Issue tracker of Emacs+, Spacemacs and Doom Emacs are full of issues caused by invalid environment. 

I am not sure how it works in Emacs Mac Port, but they have so many patches for MacOS that it's not GNU Emacs anymore.

I have few questions though:

1. How do you start Emacs? I suspect that via Finder/Spotlight by opening Emacs.app.
2. How do you set your default shell? By using `chsh` or by changing configuration of your terminal emulator, for example Terminal.app? The fact that you get `zsh` tells me that something is fishy (not related to `fish`) happening here.

**Comment by lakkiy** [comment]
> Hey @404cn,
> 
> Environment variables is always a burden on macos and is a constant topic of confusion and troubles. Issue tracker of Emacs+, Spacemacs and Doom Emacs are full of issues caused by invalid environment.
> 
> I am not sure how it works in Emacs Mac Port, but they have so many patches for MacOS that it's not GNU Emacs anymore.
> 
> I have few questions though:
> 
>     1. How do you start Emacs? I suspect that via Finder/Spotlight by opening Emacs.app.
> 
>     2. How do you set your default shell? By using `chsh` or by changing configuration of your terminal emulator, for example Terminal.app? The fact that you get `zsh` tells me that something is fishy (not related to `fish`) happening here.

Thanks for your reply.

1. Yeah I start emacs from Spotlight, it's fine to start emacs from terminal emulator, I should wrote that before, sorry.
2. I changed my default shell like fish doc [how do I make fish my default shell](https://fishshell.com/docs/current/faq.html#how-do-i-make-fish-my-default-shell), actually by this two commands:

```sh
echo /usr/local/bin/fish | sudo tee -a /etc/shells

chsh -s /usr/local/bin/fish
```

**Comment by lakkiy** [comment]
And for me this is not an urgent problem, but I just want emacs to be used intuitively.

**Comment by d12frosted** [comment]
@404cn 🤔 It's strange. Cause even without any configurations (e.g. when I wipe out my `init.el` and start Emacs.app from Finder), I still get `fish` as my shell via `M-; (get-env "SHELL") <ret>`. I even specifically tested with https://emacsformacos.com/ to verify that no custom rule I might have in scripts that build Emacs interfere with this behaviour. 

Just a random though. What do you see if you open System Preferences > Users & Groups > /unlock more features via clicking the lock icon in the left bottom corner/ > /right click on your user/ > Advanced options? What is your login shell there?

**Comment by lakkiy** [comment, marked as answer]
@d12frosted

Yes it's `/use/local/bin/fish` and It works after I reboot my system.

**Reply by d12frosted** [reply]
I converted it into discussion 😸 Feel free to open more threads 💯

**Comment by d12frosted** [comment]
Right. After changing your **login** shell you need to restart or log-out and log-in 😂 Does it work now?
